src/main/java/meteordevelopment/meteorclient/gui/utils/SettingsWidgetFactory.java

```diff
@@ -5,10 +5,49 @@
 
 package meteordevelopment.meteorclient.gui.utils;
 
+import meteordevelopment.meteorclient.gui.DefaultSettingsWidgetFactory;
 import meteordevelopment.meteorclient.gui.GuiTheme;
 import meteordevelopment.meteorclient.gui.widgets.WWidget;
+import meteordevelopment.meteorclient.gui.widgets.containers.WTable;
+import meteordevelopment.meteorclient.settings.Setting;
 import meteordevelopment.meteorclient.settings.Settings;
 
-public interface SettingsWidgetFactory {
-    WWidget create(GuiTheme theme, Settings settings, String filter);
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+public abstract class SettingsWidgetFactory {
+    private static final Map<Class<?>, Function<GuiTheme, DefaultSettingsWidgetFactory.Factory>> customFactories = new HashMap<>();
+
+    protected final GuiTheme theme;
+    protected final Map<Class<?>, Factory> factories = new HashMap<>();
+
+    public SettingsWidgetFactory(GuiTheme theme) {
+        this.theme = theme;
+
+        registerDefaultFactories();
+    }
+
+    /** {@code SettingsWidgetFactory.registerCustomFactory(SomeSetting.class, (theme) -> (table, setting) -> {//create widget})} */
+    public static void registerCustomFactory(Class<?> settingClass, Function<GuiTheme, Factory> factoryFunction) {
+        customFactories.put(settingClass, factoryFunction);
+    }
+
+    public static void unregisterCustomFactory(Class<?> settingClass) {
+        customFactories.remove(settingClass);
+    }
+
+    protected abstract void registerDefaultFactories();
```

Comment:
Why is this an abstract method? Just register them in the constructor of DefaultSettingsWidgetFactory

Reply:
i originally added it so that it was possible to extend `DefaultSettingsWidgetFactory` while also modifying the factories themselves, but then i moved all the non-factory code out of it and forgot to remove that lol